### DISCUSSION TITLE
Proper usage of YAML configuration file (pruning, freezing, etc.)

### DISCUSSION THREAD

**Discussion by aliasboink** [yoshitomo-matsubara/torchdistill#227] [Q&A] [answered]
Hello again! 

I am now trying to compress some data on the DenseDeepGCN and I'm quite unsure on how to exactly pinpoint what I wanna do through the YAML file. Firstly I will present what I wanna do and secondly the difficulties I have.

I want to add an `EntropyBottleneckLayer` after the `KNN dilation`, which means here:
https://github.com/lightaime/deep_gcns_torch/blob/master/examples/sem_seg_dense/architecture.py#L49-L50

After adding said bottleneck, I wanna train it. The way in which I wanna do it is in like Fig. 3 (https://arxiv.org/abs/2011.12913), effectively leaving the rest of the model as is and retraining the **encoder** (head as previously called) section.

Now, I've tried figuring out how to implement this through reading the paper and reading examples (such as [this](https://github.com/yoshitomo-matsubara/sc2-benchmark/blob/main/configs/pascal_voc2012/supervised_compression/entropic_student/deeplabv3_splittable_resnet50-fp-beta1.28_from_deeplabv3_resnet50.yaml), but I can really only make educated guesses). 

What I am asking is effectively: how do I prune the model and selectively use the forward hook manager properly in case I want to experiment with more configuration? (i.e., choose which layers train which, Fig 1. (b) would be an example of that)

Thank you a lot!

P.S.: I saw the lil mention of the dictionary that it can be empty and effectively ignored, whoops! (referring to my older questions).

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @banciuadrian 

To prune models like using only the first layers of a model at some training stage, you can specify `sequential` field for teacher and student models like [this](https://github.com/yoshitomo-matsubara/sc2-benchmark/blob/main/configs/ilsvrc2012/supervised_compression/entropic_student/splitable_resnet50-fp-beta5.12_from_resnet50.yaml#L87-L103).

In the example, it has two training stages. At the first training stage, only the first layers of teacher and student models are executed:
- The pruned teacher model:
  - `nn.Sequential` consisting of `['conv1', 'bn1', 'relu', 'maxpool', 'layer1', 'layer2', 'layer3', 'layer4']`
  - The whole model is frozen by `requires_grad: False`
  - Extract output of `['layer1', 'layer2', 'layer3', 'layer4']`
- The pruned student model:
  - `nn.Sequential` consisting of `['bottleneck_layer', 'layer2', 'layer3', 'layer4']`
  - Only `['layer2', 'layer3', 'layer4']` are frozen
  - Extract output of `['bottleneck_layer', 'layer2', 'layer3', 'layer4', 'bottleneck_layer.entropy_bottleneck']`

The extracted outputs are used to compute losses defined with the module paths like `bottleneck_layer`. In the example, [five different loss values are linearly combined](https://github.com/yoshitomo-matsubara/sc2-benchmark/blob/main/configs/ilsvrc2012/supervised_compression/entropic_student/splitable_resnet50-fp-beta5.12_from_resnet50.yaml#L117-L187).

Note that at the validation session and at the second training stage, the full models are used.

**Reply by aliasboink** [reply]
Hi @yoshitomo-matsubara 

Based on syntax, I assume this just passes the input in sequential fashion through the given layers. I assume this wouldn't really work if we have something like we do in DenseDeepGCN:
```python
feats = [self.head(inputs, self.knn(inputs[:, 0:3]))]
```
as this is the first thing in the `forward()` function. This doesn't simply take the input and pass it through the layer.

I am also guessing this doesn't really work with the rest of the forward function:
```python
for i in range(self.n_blocks-1):
    feats.append(self.backbone[i](feats[-1]))
feats = torch.cat(feats, dim=1)

fusion = torch.max_pool2d(self.fusion_block(feats), kernel_size=[feats.shape[2], feats.shape[3]])
fusion = torch.repeat_interleave(fusion, repeats=feats.shape[2], dim=2)
return self.prediction(torch.cat((fusion, feats), dim=1)).squeeze(-1)
 ```
This calls functions from `torch` and the backbone does a lot of appending, which is not done in the `layer`, but the `forward` function. Does this mean the pruning only works in the cases in which the model can be represented as a sequential set of layers?

Am I wrong to assume this?

**Reply by yoshitomo-matsubara** [reply]
Hi @banciuadrian 

> Does this mean the pruning only works in the cases in which the model can be represented as a sequential set of layers

Partially yes. If you cannot represent your pruned module as `nn.Sequential`, `sequential` entry in the yaml won't help.

Instead, you may want to register your own "special module" and define it in your yaml file. e.g., [`HeadRCNN`](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L324-L344) used in [this yaml config](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/single_stage/ghnd/custom_fasterrcnn_resnet50_fpn_from_fasterrcnn_resnet50_fpn.yaml#L70-L98).

If `feats = [self.head(inputs, self.knn(inputs[:, 0:3]))]` is all you need from your model, probably your special module looks something like this:
```python
@register_special_module
class HeadDenseDeepGCN(SpecialModule):
    def __init__(self, head_rcnn, **kwargs):
        super().__init__()
        tmp_ref_model = kwargs.get('teacher_model', None)
        ref_model = kwargs.get('student_model', tmp_ref_model)
        if ref_model is None:
            raise ValueError('Either student_model or teacher_model has to be given.')

        self.knn = ref_model.knn
        self.head = ref_model.head

def forward(self, inputs):
        feats = [self.head(inputs, self.knn(inputs[:, 0:3]))]
        return self.seq(images.tensors)
```
